### PR TITLE
Convert to new override syntax

### PR DESCRIPTION
**Pull request by kraj**

### CODE REVIEW

**Comment by otavio** [comment]
@kraj, thanks for looking into it. How did you validate it?

**Comment by kraj** [comment]
> @kraj, thanks for looking into it. How did you validate it?

I have one freescale imx6ull based device which I booted with it and also world builds ensured that no evident fallouts are there. but ofcourse its something BSP maintainers should test out on available hardware.

File: recipes-bsp/imx-mkimage/imx-boot_1.0.bb
Comment (zandrey) [inline recipes-bsp/imx-mkimage/imx-boot_1.0.bb:74]: This (and consecutive substitutions) should not be made, as those functions are called by `do_compile()` and `do_deploy()`.

Changing the name here would lead to failures in executing `do_compile` and `do_deploy`.

I would suggest this substitution to be reverted for those internal functions.

**Review by zandrey — CHANGES_REQUESTED** [review]

File: recipes-bsp/imx-mkimage/imx-boot_1.0.bb
Comment (thochstein) [reply recipes-bsp/imx-mkimage/imx-boot_1.0.bb:74]: Good catch. @kraj can you do a build for an i.MX 8 and 8M machine, perhaps `imx8qm-mek` and `imx8mp-lpddr4-evk`? Maybe just a parse is all that's needed to test for this sort of problem?

Comment (zandrey) [reply recipes-bsp/imx-mkimage/imx-boot_1.0.bb]: @thochstein I actually just did it for `imx8mp-lpddr4-evk`, and parsing was only breaking down in `imx-boot` recipe.

-- andrey

**Review by zandrey — APPROVED** [review]

File: recipes-bsp/imx-mkimage/imx-boot_1.0.bb
Comment (kraj) [reply recipes-bsp/imx-mkimage/imx-boot_1.0.bb]: @thochstein if I have time and resources I would do it but I dont sadly, so everyone has to do their bit, there are many machines in this layer, I cant test them all.

**Comment by thochstein** [comment]
I noticed overrides beginning with `_imx`. Grepping for all such I also see some `_virtual`. Is there a mechanism so these can be flagged as errors?

```
recipes-bsp/libimxdmabuffer/libimxdmabuffer_1.0.1.bb:LIBG2D_PACKAGECONFIG_imxdpu = ""
recipes-multimedia/gstreamer/gstreamer1.0-plugins-imx_2.0.0.bb:LIBG2D_DPU_OPTION_imxdpu = "-Dg2d-based-on-dpu=true"
recipes-multimedia/gstreamer/gstreamer1.0-plugins-imx_2.0.0.bb:LIBG2D_DEPENDENCIES_imxdpu = "virtual/libg2d libdrm"
classes/use-imx-headers.bbclass:DEPENDS:append_imx = " linux-imx-headers"
classes/use-imx-headers.bbclass:PACKAGE_ARCH_imx ?= "${MACHINE_SOCARCH}"
conf/machine/include/imx-base.inc:MACHINE_SOCARCH_FILTER:append_imx = " \
conf/machine/include/imx-base.inc:PREFERRED_PROVIDER_virtual/egl_imxgpu        ?= "imx-gpu-viv"
conf/machine/include/imx-base.inc:PREFERRED_PROVIDER_virtual/libgl_imxgpu3d    ?= "imx-gpu-viv"
conf/machine/include/imx-base.inc:PREFERRED_PROVIDER_virtual/libgles1_imxgpu3d ?= "imx-gpu-viv"
conf/machine/include/imx-base.inc:PREFERRED_PROVIDER_virtual/libgles2_imxgpu3d ?= "imx-gpu-viv"
conf/machine/include/imx-base.inc:PREFERRED_PROVIDER_virtual/libg2d_imxdpu     ?= "imx-dpu-g2d"
conf/machine/include/imx-base.inc:PREFERRED_VERSION_weston_imx ?= "9.0.0.imx"
conf/machine/include/imx-base.inc:# Add the ability to specify _imx machines
recipes-graphics/imx-gpu-viv/imx-gpu-viv-6.inc:DEPENDS:append_imxdrm = " libdrm wayland"
recipes-graphics/wayland/weston-init.bbappend:INI_UNCOMMENT_ASSIGNMENTS:append_imx = " \
recipes-graphics/xorg-xserver/xserver-xorg_%.bbappend:OPENGL_PKGCONFIGS:remove_imxdrm     = "dri glx"
```

**Comment by zandrey** [comment]
@kraj @otavio 

Can anyone of you pick up the commit 86b6c13004bd9c855d0dbbdc4bfd127e08876886 from my fork and either include it in this PR or cherry-pick it separately once the PR is merged?

This solves the mainline U-Boot build issue when Community BSP flavor is used with new overrides.

-- andrey

**Comment by angolini** [comment]
@zandrey, which board you have built with your PR? I have this on my local dir, but I'm still facing some issues on fslc distro for imx8mm-ddr4-evk board

(but yes, your commit is needed)

**Comment by angolini** [comment]
I have tested this PR using 

```
MACHINE = 'imx8mm-lpddr4-evk'
DISTRO ?= 'fsl-framebuffer'
```

and core-image-minimal, and the board is booting
I have built also the same for `imx8mm-ddr4-evk` and i

**Comment by otavio** [comment]
@kraj can you cherry-pick @zandrey fix? so we have a single PR fixing all issues at once?

**Comment by zandrey** [comment]
> 
> 
> @zandrey, which board you have built with your PR?

@angolini I'm building the `fslc-wayland` distro for `imx8mp-lpddr4-evk` machine. I've picked up patches from @kraj locally for this, hence I asked for my commit to be cherry-picked on top of this PR.

-- andrey

**Comment by kraj** [comment]
> @kraj @otavio
> 
> Can anyone of you pick up the commit [86b6c13](https://github.com/Freescale/meta-freescale/commit/86b6c13004bd9c855d0dbbdc4bfd127e08876886) from my fork and either include it in this PR or cherry-pick it separately once the PR is merged?
> 
> This solves the mainline U-Boot build issue when Community BSP flavor is used with new overrides.
> 
> -- andrey

thanks I have cherry-picked it.

File: recipes-support/opencv/opencv_4.5.2.imx.bb
Comment (dankm) [inline recipes-support/opencv/opencv_4.5.2.imx.bb:252]: Looks like these aren't overrides. It should remain SRCREV_extra.

**Comment by thochstein** [comment]
I updated the problems I found and did a PR for the source of this PR:

https://github.com/YoeDistro/meta-freescale/pull/1

I also included a patch to undo the `SRCREV_extra` problem.

File: conf/machine/include/imx-base.inc
Comment (texierp) [inline conf/machine/include/imx-base.inc:72]: Are you sure @thochstein ? I think that we need to keep the original syntax -> `PREFERRED_PROVIDER_` here

Comment (thochstein) [reply conf/machine/include/imx-base.inc:72]: Yes, I believe you're right. Will fix.

**Comment by otavio** [comment]
@kraj can you rebase this and handle the @thochstein PRs he sent to Yoe's repo? I'd like to run some testing again and prepare for merging.

File: conf/machine/include/imx-base.inc
Comment (kraj) [reply conf/machine/include/imx-base.inc]: done, I have squashed this one into original commit from pull 1

**Comment by kraj** [comment]
> @kraj can you rebase this and handle the @thochstein PRs he sent to Yoe's repo? I'd like to run some testing again and prepare for merging.

done

File: conf/machine/include/imx-base.inc
Comment (thochstein) [inline conf/machine/include/imx-base.inc:275]: This one snuck through...

**Comment by thochstein** [comment]
Seems that `MACHINEOVERRIDES_EXTENDER` is not working, it doesn't make it into `MACHINEOVERRIDES`.

```
$ bitbake u-boot-imx -e | grep ^MACHINEOVERRIDES_EXTENDER=
MACHINEOVERRIDES_EXTENDER="imxdrm:imxvpu:imxgpu:imxgpu2d:imxgpu3d"
$ bitbake u-boot-imx -e | grep ^MACHINEOVERRIDES=
MACHINEOVERRIDES="aarch64:armv8a:imx:use-nxp-bsp:imx-boot-container:mx8:mx8m:mx8mm:imx8mm-lpddr4-evk"
```

**Comment by thochstein** [comment]
Fixed `_mx27` and `MACHINEOVERRIDES_EXTENDER`: https://github.com/YoeDistro/meta-freescale/pull/3

File: classes/image_types_fsl.bbclass
Comment (otavio) [inline classes/image_types_fsl.bbclass:22]: Seems broken. @twoerner see if using `do_image:uboot_mxsboot_nand` here and other places might help.

Comment (kraj) [reply classes/image_types_fsl.bbclass]: ah yes confused script, I pushed a fix in branch

**Comment by twoerner** [comment]
this is my diff. with this diff building core-image-base for imx233-olinuxino-maxi works


```
diff --git a/classes/image_types_fsl.bbclass b/classes/image_types_fsl.bbclass
index 9f76273f..cdd16a66 100644
--- a/classes/image_types_fsl.bbclass
+++ b/classes/image_types_fsl.bbclass
@@ -9,7 +9,7 @@ UBOOT_SUFFIX ?= "bin"
 MXSBOOT_NAND_ARGS ?= ""
 
 # U-Boot mxsboot generation for uSD
-do_image_uboot:mxsboot_sdcard[depends] += "u-boot-mxsboot-native:do_populate_sysroot \
+do_image_uboot_mxsboot_sdcard[depends] += "u-boot-mxsboot-native:do_populate_sysroot \
                                            u-boot:do_deploy"
 IMAGE_CMD:uboot-mxsboot-sdcard() {
     mxsboot sd ${DEPLOY_DIR_IMAGE}/u-boot-${MACHINE}.${UBOOT_SUFFIX} \
@@ -19,7 +19,7 @@ IMAGE_CMD:uboot-mxsboot-sdcard() {
 }
 
 # U-Boot mxsboot generation for NAND
-do_image_uboot:mxsboot_nand[depends] += "u-boot-mxsboot-native:do_populate_sysroot \
+do_image_uboot_mxsboot_nand[depends] += "u-boot-mxsboot-native:do_populate_sysroot \
                                          u-boot:do_deploy"
 IMAGE_CMD:uboot-mxsboot-nand() {
     mxsboot ${MXSBOOT_NAND_ARGS} nand \
@@ -33,7 +33,7 @@ IMAGE_CMD:uboot-mxsboot-nand() {
 # image stream built before the wic generation
 do_image_wic[depends] += " \
     ${@bb.utils.contains('IMAGE_FSTYPES', 'uboot-mxsboot-sdcard', \
-                                          '${IMAGE_BASENAME}:do_image_uboot:mxsboot_sdcard', '', d)} \
+                                          '${IMAGE_BASENAME}:do_image_uboot_mxsboot_sdcard', '', d)} \
 "
 
 # We need to apply a fixup inside of the partition table
```

**Comment by twoerner** [comment]
i think @kraj's lastest patch covers it, thanks!

**Review by otavio — APPROVED** [review]

**Comment by twoerner** [comment]
> Folks, let's do another test cycle prior merging? it seems this is good to go now.

Okay, I can build test (core-image-base) for:
- imx233-olinuxino-maxi
- imx233-olinuxino-micro
- imx233-olinuxino-mini
- imx233-olinuxino-nano
- imx28evk
- imx53qsb
- wandboard

and runtime test for:
- imx233-olinuxino-maxi
- imx28evk
- imx53qsb
- wandboard

**Comment by angolini** [comment]
I can build and runtime:
* imx8mm-lpddr4-evk
* imx6ullevk
* imx7dsabresd (fingers crossed)

I can probably build something else if needed overnight

I expect to finish probably only on Saturday

**Comment by texierp** [comment]
I can build and runtime:

 - imx7s-warp
 - imx7d-pico

**Comment by thochstein** [comment]
I can build and runtime:

- imx6qdlsabresd
- imx8mq-evk

**Comment by texierp** [comment]
JFYI, when I try to build for both `imx7s-warp` and `imx7d-pico` (with [1]), I get the following build error:

```
...
|   ./tools/mkimage -n spl/u-boot-spl.cfgout -T imximage -e 0x00912000 -d spl/u-boot-spl.bin SPL >SPL.log  && cat SPL.log
| Image Type:   Freescale IMX Boot Image
| Image Ver:    2 (i.MX53/6/7 compatible)
| Mode:         DCD
| Data Size:    45056 Bytes = 44.00 KiB = 0.04 MiB
| Load Address: 00911420
| Entry Point:  00912000
| make[1]: Leaving directory '/build/tmp/work/imx7d_pico-poky-linux-gnueabi/u-boot-fslc/v2021.07+gitAUTOINC+587e796995-r0/build/pico-dwarf-imx7d_defconfig'
| make: Leaving directory '/build/tmp/work/imx7d_pico-poky-linux-gnueabi/u-boot-fslc/v2021.07+gitAUTOINC+587e796995-r0/git'
| /build/tmp/work/imx7d_pico-poky-linux-gnueabi/u-boot-fslc/v2021.07+gitAUTOINC+587e796995-r0/temp/run.do_compile.1244816: 187: /build/tmp/work/imx7d_pico-poky-linux-gnueabi/u-boot-fslc/v2021.07+gitAUTOINC+587e796995-r0/temp/run.do_compile.1244816: Bad substitution

```

Where line 187 is:

```
cp /build/tmp/work/imx7s_warp-poky-linux-gnueabi/u-boot-fslc/v2021.07+gitAUTOINC+587e796995-r0/build/${config}/${binary} /build/tmp/work/imx7s_warp-poky-linux-gnueabi/u-boot-fslc/v2021.07+gitAUTOINC+587e796995-r0/build/${config}/u-boot-${type}.${UBOOT_SUFFIX:pn-u-boot-fslc}
```

[1] - https://github.com/texierp/pr_yoe

**Comment by kraj** [comment]
@texierp I can build it fine here with yoe

```
Build Configuration:
BB_VERSION           = "1.51.1"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "arm-yoe-linux-gnueabi"
MACHINE              = "imx7d-pico"
DISTRO               = "yoe"
DISTRO_VERSION       = "2021.07"
TUNE_FEATURES        = "arm armv7ve vfp thumb neon callconvention-hard"
TARGET_FPU           = "hard"
....

bitbake virtual/bootloader

```

**Comment by twoerner** [comment]
> > Folks, let's do another test cycle prior merging? it seems this is good to go now.
> 
> Okay, I can build test (core-image-base) for:
> 
> * imx233-olinuxino-maxi
> * imx233-olinuxino-micro
> * imx233-olinuxino-mini
> * imx233-olinuxino-nano
> * imx28evk
> * imx53qsb
> * wandboard

The above all build test fine (core-image-base, OE-only, nodistro)

> and runtime test for:
> 
> * imx233-olinuxino-maxi
> * imx28evk
> * imx53qsb
> * wandboard

and they all ran fine.

**Comment by otavio** [comment]
> JFYI, when I try to build for both `imx7s-warp` and `imx7d-pico` (with [1]), I get the following build error:
...
> [1] - https://github.com/texierp/pr_yoe

I wonder if this is reproducible if running against `hardknoot`. I am puzzled as @kraj couldn't reproduce it, so that it might be due to something related to the host?

**Comment by texierp** [comment]
> > JFYI, when I try to build for both `imx7s-warp` and `imx7d-pico` (with [1]), I get the following build error:
> > ...
> > [1] - https://github.com/texierp/pr_yoe
> 
> I wonder if this is reproducible if running against `hardknoot`. I am puzzled as @kraj couldn't reproduce it, so that it might be due to something related to the host?

Thanks for the feedback @otavio 

Just check, and the build is fine with `hardknott` ;). Regarding the host, I use a docker container (with `kas`, in [1]).
I will investigate but my problem seems related to [2] (bad substitution).

Anyway, since the build is fine for @kraj , you can merge this PR as is.

[1] - https://github.com/texierp/pr_yoe/blob/master/kas-container
[2] - https://github.com/Freescale/meta-freescale/blob/791726c7bb10d94e517fe0423dc8cb3689424b78/conf/machine/include/imx-base.inc#L64

**Comment by otavio** [comment]
I've merged this as this moves things on the right direction. 
We can address regresions on new PRs.

**Comment by angolini** [comment]
> I can build and runtime:
> 
>     * imx8mm-lpddr4-evk
> 
>     * imx6ullevk
> 
>     * imx7dsabresd (fingers crossed)
> 
> 
> I can probably build something else if needed overnight
> 
> I expect to finish probably only on Saturday

* imx6ullevk = OK

**Pull request merged**
